Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp

```diff
@@ -142,172 +151,222 @@ void wifi_reset() {
 }
 
 void mount_file_sys(uint8_t disk_type) {
-  if (disk_type == FILE_SYS_SD) {
+  if (disk_type == FILE_SYS_SD)
     TERN_(SDSUPPORT, card.mount());
-  }
   else if (disk_type == FILE_SYS_USB) {
   }
 }
 
-static void dma_init() {
-  #if 0
-  __HAL_RCC_DMA1_CLK_ENABLE();
-
-  //HAL_NVIC_SetPriorityGrouping(NVIC_PRIORITYGROUP_0);
-  HAL_NVIC_SetPriority(DMA1_Channel5_IRQn, 4, 0);
-  HAL_NVIC_EnableIRQ(DMA1_Channel5_IRQn);
-
-  hdma_usart1_rx.Instance = DMA1_Channel5;
-  //hdma_usart1_rx.Init.Channel = DMA_CHANNEL_4;
-  hdma_usart1_rx.Init.Direction = DMA_PERIPH_TO_MEMORY;
-  hdma_usart1_rx.Init.PeriphInc = DMA_PINC_DISABLE;
-  hdma_usart1_rx.Init.MemInc = DMA_MINC_ENABLE;
-  hdma_usart1_rx.Init.PeriphDataAlignment = DMA_PDATAALIGN_BYTE;
-  hdma_usart1_rx.Init.MemDataAlignment = DMA_PDATAALIGN_BYTE;
-  hdma_usart1_rx.Init.Mode = DMA_NORMAL;
-  hdma_usart1_rx.Init.Priority = DMA_PRIORITY_VERY_HIGH;
-  if (HAL_DMA_Init((DMA_HandleTypeDef *)&hdma_usart1_rx) != HAL_OK) {
-    Error_Handler();
+#include <libmaple/timer.h>
+#include <libmaple/util.h>
+#include <libmaple/rcc.h>
+
+#include <boards.h>
+#include <wirish.h>
+
+#include <libmaple/dma.h>
+#include <libmaple/bitband.h>
+
+#include <libmaple/libmaple.h>
+#include <libmaple/gpio.h>
+#include <libmaple/usart.h>
+#include <libmaple/ring_buffer.h>
+
+void exchangeFlashMode(char dmaMode) {
+  if (flash_dma_mode != dmaMode) {
+    flash_dma_mode = dmaMode;
+    if (flash_dma_mode == 1) {
+      // uint8_t buf[2];
+      // W25QXX.SPI_FLASH_BufferRead(buf, 0, 2);
+      // ZERO(buf);
+      // SPI_TFT.tftio.WriteSequence((uint16_t*)buf, 1);
+    }
+    else {
+      // spi_tx_dma_disable(SPI2);
+      // spi_rx_dma_disable(SPI2);
+      dma_disable(DMA1, DMA_CH5);
+      //dma_disable(DMA1, DMA_CH4);
+      dma_clear_isr_bits(DMA1, DMA_CH4);
+      //dma_clear_isr_bits(DMA1, DMA_CH5);
+
+      // dma_disable(DMA1, DMA_CH3);
+      // dma_disable(DMA1, DMA_CH2);
+      // dma_clear_isr_bits(DMA1, DMA_CH3);
+      // dma_clear_isr_bits(DMA1, DMA_CH2);
+    }
   }
+}
 
+static bool longName2DosName(const char* longName, uint8_t* dosName) {
+  uint8_t i = 11;
+  while (i) dosName[--i] = '\0';
+  while (*longName) {
+    uint8_t c = *longName++;
+    if (c == '.') {                   // For a dot...
+  if (i == 0) { return false; }
+  else { strcat((char *)dosName,".GCO"); return dosName[0] != '\0'; }
+    }
+    else {
+      // Fail for illegal characters
+      PGM_P p = PSTR("|<>^+=?/[];,*\"\\");
+      while (uint8_t b = pgm_read_byte(p++)) if (b == c) return false;
+      if (c < 0x21 || c == 0x7F) return false;           // Check size, non-printable characters
+      dosName[i++] = (c < 'a' || c > 'z') ? (c) : (c + ('A' - 'a')); // Uppercase required for 8.3 name
+    }
+    if (i >= 5) { strcat((char *)dosName,"~1.GCO"); return dosName[0] != '\0'; }
```

Comment:
This does not look robust or correct.

What if you have two long filenames which share the same leading 5 characters?

Reply:
The old file will be deleted and replaced with the new one. I tested it.